--- a/glib/gslist.c
+++ b/glib/gslist.c
@@ -146,7 +146,13 @@
 g_slist_free_full (GSList         *list,
 		   GDestroyNotify  free_func)
 {
-  g_slist_foreach (list, (GFunc) free_func, NULL);
+  GSList *node = list;
+  while (node != NULL)
+    {
+      GSList *next = node->next;
+      free_func (node->data);
+      node = next;
+    }
   g_slist_free (list);
 }
 
--- a/glib/glist.c
+++ b/glib/glist.c
@@ -159,7 +159,13 @@
 g_list_free_full (GList          *list,
                   GDestroyNotify  free_func)
 {
-  g_list_foreach (list, (GFunc) free_func, NULL);
+  GList *node = list;
+  while (node != NULL)
+    {
+      GList *next = node->next;
+      free_func (node->data);
+      node = next;
+    }
   g_list_free (list);
 }
 
